File: LC/Python/Medium/271.py
class Codec:
    def encode(self, strs: List[str]) -> str:
        """Encodes a list of strings to a single string.
        """
        encoded = []
        
        for word in strs:
            encoded.append(f"{len(word)}${word}")
        
        return "".join(encoded)
        
        

    def decode(self, s: str) -> List[str]:
        """Decodes a single string to a list of strings.
        """
        ans = []
        i = 0
        
        while i < len(s):
            j = i
            while j < len(s) and s[j] != "$":
                j += 1
            curr_len = int(s[i:j])
            
            ans.append(s[j+1: j + 1 + curr_len])
            i = j + 1 + curr_len
        
        return ans
        

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.decode(codec.encode(strs))

File: LC/C++/Medium/271.cpp
class Codec {
public:

    // Encodes a list of strings to a single string.
    string encode(vector<string>& strs) {
        std::string ans{};

        for (auto& str : strs) {
            ans += to_string(str.size());
            ans += "$";
            ans += str;
        }

        return ans;
    }

    // Decodes a single string to a list of strings.
    vector<string> decode(string s) {
        std::vector<string> ans{};
        int i = 0;

        while (i < s.size()) {
            string len;

            while (isdigit(s[i])) {
                len += s[i++];
            }

            i++;

            int length = stoi(len);
            ans.push_back(s.substr(i, length));
            i += length;
        }

        return ans;
    }
};

// Your Codec object will be instantiated and called as such:
// Codec codec;
// codec.decode(codec.encode(strs));